Adding decoupled support

pauarge opened the pull request:
## Description

This is the initial iteration in order to add support for decoupled WordPress architectures in the plugin. In this PR, we're extending the `post` and `page` models in the WordPress REST API to include the `parsely-meta` field. This field returns a JSON object containing the meta tags using the same function that's being used to generate the data in the front-end, `construct_parsely_metadata`.

This behavior is controllable by the `wp_parsely_enable_rest_api_support` filter.

## Motivation and Context

See https://github.com/Parsely/wp-parsely/issues/487

## How Has This Been Tested?

In a WordPress installation with wp-parsely, navigate to `?rest_route=/wp/v2/posts/1` and you should expect to see the `parsely-meta` object.

## Screenshots (if appropriate):

<img width="621" alt="Screen Shot 2021-11-18 at 4 29 46 PM" src="https://user-images.githubusercontent.com/7188409/142445338-08467285-7f97-41e2-8518-7f1137a63d32.png">

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

pauarge committed:
Starting with REST API

pauarge committed:
Returning metadata on REST

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Adding testing class

pauarge committed:
Test rest_api_init

pauarge committed:
Adding filter to disable REST API support

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Adding registration test

pauarge committed:
Check for get_callback function

pauarge committed:
Checking other functions are not defined

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Adding explicit return type on parsely meta

pauarge committed:
Merge branch 'develop' into add/decoupled

GaryJones reviewed (approved):
Works equally well for page:

<img width="306" alt="Screenshot 2021-11-25 at 11 42 31" src="https://user-images.githubusercontent.com/88371/143435689-da1f1b90-a56b-4217-98f4-e02deb115ced.png">

✅ 

Any reason this couldn't be pushed to 3.0? New useful feature for users (developers).

GaryJones reviewed (changes requested):
See https://github.com/Parsely/wp-parsely/pull/500 for some discussion about a different format.

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Merge branch 'develop' into add/decoupled

chriszarate commented on src/class-rest.php:
```diff
@@ -0,0 +1,78 @@
+		};
+
+		$args = array( 'get_callback' => $callback );
+		register_rest_field( 'post', self::PARSELY_META_REST_FIELD_NAME, $args );
+		register_rest_field( 'page', self::PARSELY_META_REST_FIELD_NAME, $args );
```
Could we:

1. Target all public post types that support `editor`? This would allow easy custom post type support.
2. Allow the post types to be filtered before registering the field?

chriszarate commented:
@pauarge I've created a draft PR of our decoupled plugin bundle against this branch, which extracts the data so that it can be exposed by GraphQL:

https://github.com/Automattic/vip-decoupled-bundle/pull/36

In order to extract the data we need, we must use `construct_parsely_metadata` from the `$parsely` global. Is this stable / safe?

chriszarate commented:
@pauarge Would you be open to conditionally registering WPGraphQL fields here as well? Lots of plugins take this approach, behind a simple `function_exists` check.

pauarge committed:
Nesting fields in array

pauarge commented:
Thanks for your comments @chriszarate !

> Target all public post types that support editor? This would allow easy custom post type support.
    Allow the post types to be filtered before registering the field?

I've updated the code to use `track_post_types` and `track_page_types` from the plugin's settings in wp-admin. This way we let the user customize whatever they want to track and match the behavior of the WP FE code insertion.

> GraphQL

We've had some discussions with @GaryJones regarding the addition of GraphQL to the plugin. We thought about doing so in the next version. From our side, we're more than happy to integrate the code that you made to the main plugin.

> In order to extract the data we need, we must use construct_parsely_metadata from the $parsely global. Is this stable / safe?

Yes

pauarge committed:
Using options to select object_type

GaryJones commented on src/class-rest.php:
```diff
@@ -18,7 +18,8 @@
 class Rest {
-	private const PARSELY_META_REST_FIELD_NAME = 'parsely-meta';
+	private const PARSELY_REST_VERSION         = '1.0.0';
+	private const PARSELY_META_REST_FIELD_NAME = 'parsely';
```
The `META` part of this constant name could be removed, as it's the name for the whole field.

GaryJones commented on src/class-rest.php:
```diff
@@ -68,11 +69,18 @@ public function register_parsely_meta(): void {
-		register_rest_field( 'post', self::PARSELY_META_REST_FIELD_NAME, $args );
-		register_rest_field( 'page', self::PARSELY_META_REST_FIELD_NAME, $args );
+		$options     = $this->parsely->get_options();
+		$object_type = array_merge( $options['track_post_types'], $options['track_page_types'] );
```
We could probably do with an `array_unique()` here, in case someone has registered the same type for both post and page (we don't block the checkboxes yet).

Also, should there be a filter here?

pauarge replied:
> We could probably do with an array_unique() here

Agreed, adding in a commit.

> Also, should there be a filter here?

We could, I imagine that you are suggesting this in case a user wants logic more complex than ON/OFF, which the checkboxes already provide? I would add the filter in a further iteration to keep it simple at the beginning.

GaryJones replied:
More that I'm thinking an engineer may want to guarantee what REST API outputs are affected, regardless of what some colleague may fiddle with the checkboxes are accidentally break things.

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Renaming field

pauarge committed:
Adding object type filter

pauarge committed:
Merge branch 'develop' into add/decoupled

GaryJones commented on src/class-rest.php:
```diff
@@ -79,8 +79,17 @@ public function register_parsely_meta(): void {
+		 *
+		 * @param string[] $object_type Array of strings containing the object types, i.e. `page`, `post`, `term`.
+		 */
+		$object_type = apply_filters( 'wp_parsely_rest_object_type', $object_type );
```
Let's make `$object_type` and the filter name plural, to indicate that we're expected an array or more than one type (at least by default).

pauarge replied:
I did use singular because `register_rest_field` uses singular. Then again, in that case it can be a string or an array, in our case, if should always be an array.

pauarge committed:
Changing object type to plural

pauarge committed:
Fix linting issues

pauarge committed:
Merge branch 'develop' into add/decoupled

chriszarate commented:
> We've had some discussions with @GaryJones regarding the addition of GraphQL to the plugin. We thought about doing so in the next version. From our side, we're more than happy to integrate the code that you made to the main plugin.

Sounds good. I'll work on formalizing our approach this week.

GaryJones commented on src/class-rest.php:
```diff
@@ -0,0 +1,95 @@
+		 *
+		 * @param string[] $object_types Array of strings containing the object types, i.e. `page`, `post`, `term`.
+		 */
+		$object_types = apply_filters( 'wp_parsely_rest_object_types', $object_types );
```
We need a test to demonstrate how this filter should and shouldn't be used.

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Adding filter test

pauarge committed:
Fix test

pauarge committed:
Fix linting issues

GaryJones commented on tests/Integration/RestTest.php:
```diff
@@ -0,0 +1,106 @@
+
+		self::$rest->register_parsely_meta();
+
+		$expected_fields = array( 'get_callback', 'update_callback', 'schema' );
```
Instead of checking that the array_keys of `$wp_rest_additional_fields['...']['parsely']` are these three values (which, they [always will be](https://github.com/WordPress/WordPress/blob/47c4ff7102291136473d1ae52302dd6ebfad3345/wp-includes/rest-api.php#L137) - just with different callback values, with a default of `null`), why not check that the parsely property exists for that post type in the global, and that the callback is not the default `null`?

```php
isset( $wp_rest_additional_fields['post']['parsely'] ) && ! is_null( $wp_rest_additional_fields['post']['parsely']['get_callback']
```

There's also an opportunity to check that `['get_callback']` is populated correctly - an array, with a `version` key being a string, and `meta` key with an array which is at least going to contain some expected fields for particular post types (post vs non-post).

GaryJones commented on tests/Integration/RestTest.php:
```diff
@@ -0,0 +1,106 @@
+
+		$expected_fields = array( 'get_callback', 'update_callback', 'schema' );
+
+		$this->assertEquals( $expected_fields, array_keys( $wp_rest_additional_fields['term']['parsely'] ) );
```
As well as changing what is tested for the `term`, I would also suggest checking for a lack of changes (`get_callback` is null), for a `post` or `page` type that would ordinarily get the changes.

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Adding additional tests

Cleaning up tests

Testing with dump

Testing with count

Asserting null callback

Check

Checking array key does not exist

Expecting get_callback to not be null

Remove separate process

Checking for null

Removing line

Revert changes

Check post and page don't exist

Running in separate processes

Checking for changes on post and page

Check get_callback is not null

GaryJones commented on tests/Integration/RestTest.php:
```diff
@@ -0,0 +1,125 @@
+		$expected_fields = array( 'get_callback', 'update_callback', 'schema' );
+
+		$this->assertEquals( $expected_fields, array_keys( $wp_rest_additional_fields['post']['parsely'] ) );
+		$this->assertEquals( $expected_fields, array_keys( $wp_rest_additional_fields['page']['parsely'] ) );
```
This is just testing core functionality of `register_rest_field()`. It serves no value for our tests.

GaryJones committed:
Rest: Test improvements

- Return type declarations
- Use More specific native assertions
- Consolidating repeated groups of assertions into a new helper method.
- Assert finer details about the expected get_callback value.

jblz commented on src/class-rest.php:
```diff
@@ -0,0 +1,95 @@
+	 */
+	public function register_parsely_meta(): void {
+		$callback = function( array $object ): array {
+			$post_id = $object['id'];
```
Should we check the status against [`Parsely::post_has_trackable_status`](https://github.com/Parsely/wp-parsely/blob/develop/src/class-parsely.php#L273) & return early if false?

pauarge replied:
I'm not sure about this one. The checks already happen inside `construct_parsely_metadata` and I wanted to isolate as much logic as possible from the `Rest` class.

jblz replied:
Good call.

jblz commented on src/class-rest.php:
```diff
@@ -0,0 +1,95 @@
+	 * @return void
+	 */
+	public function register_parsely_meta(): void {
+		$callback = function( array $object ): array {
```
Is there a reason this shouldn't be a class member function instead of being declared inside this one?  If we pull it out, we can cover it with a unit test.

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Extracting parsely_rest_get_callback function

GaryJones committed:
Extracting parsely_rest_get_callback function

GaryJones committed:
Rest: Further test adjustments

GaryJones commented on src/class-rest.php:
```diff
@@ -66,7 +66,7 @@ public function run(): void {
 	public function register_parsely_meta(): void {
 		$callback = function( array $object ): array {
-			$post_id = $object['id'];
+			$post_id = $object['ID'] ?? $object['id'];
```
I was finding some tests issues as `id` could not be found as a key (but `ID` could.

GaryJones commented on tests/Integration/RestTest.php:
```diff
@@ -107,16 +119,19 @@ function( $object_types ) {
 		self::assertIsString( $wp_rest_additional_fields[ $post_type ]['parsely']['get_callback']['version'] );
-		self::assertArrayHasKey( 'meta', $wp_rest_additional_fields[ $post_type ]['parsely']['get_callback'] );
+		self::assertArrayHasKey( 'meta', $wp_rest_additional_fields[ $post_type ]['parsely']['get_callback']( $object ) );
+		*/
```
With these assertions commented out, the tests should pass.

I wasn't able to get the callback to be correctly called with the right array of the WP_Post / WP_Term data, despite creating the post/page/term (category) in the main test methods.

GaryJones reviewed (changes requested):
Needs fixes, but sharing the latest work.

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Merge branch 'add/decoupled' of github.com:Parsely/wp-parsely into add/decoupled

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Adding rest callback test

Adding rest callback test

Chaning covers

Declaring parsely

Using generated metadata

Test rest callback

GaryJones commented:
Great! The `$object` arg of `assertParselyRestFieldIsConstructedCorrectly()` can be removed again, as can the code in each test method that existed to populate that arg.

pauarge committed:
Merge branch 'develop' into add/decoupled

pauarge committed:
Removing  argument

GaryJones commented on tests/Integration/RestTest.php:
```diff
@@ -115,7 +115,7 @@ function( $object_types ) {
 		$category = self::factory()->category->create( array( 'name' => 'Test Category' ) );
 		self::factory()->post->create( array( 'post_category' => array( $category ) ) );
 		$term = get_term( $category, 'post_category', 'ARRAY_A' );
```
This code can go.

GaryJones commented on tests/Integration/RestTest.php:
```diff
@@ -77,15 +77,15 @@ public function test_register_parsely_meta_registers_fields(): void {
 				'post_type' => 'page',
 			)
 		);
 		$page    = get_post( $page_id, 'ARRAY_A' );
```
This code can go.

GaryJones commented on tests/Integration/RestTest.php:
```diff
@@ -77,15 +77,15 @@ public function test_register_parsely_meta_registers_fields(): void {
 
 		$post_id = self::factory()->post->create();
 		$post    = get_post( $post_id, 'ARRAY_A' );
```
This code can go.

pauarge committed:
Removing unused code

jblz committed:
Merge branch 'develop' into add/decoupled

jblz commented on tests/Integration/RestTest.php:
```diff
@@ -0,0 +1,144 @@
+
+		add_filter(
+			'wp_parsely_rest_object_types',
+			function( $object_types ) {
```
```suggestion
			function() {
```

per: https://sonarcloud.io/project/issues?id=Parsely_wp-parsely&issues=AX2ftYPDuw8Yrhl6Lsoo&open=AX2ftYPDuw8Yrhl6Lsoo&pullRequest=489

jblz commented on src/class-rest.php:
```diff
@@ -0,0 +1,103 @@
+ *
+ * @since 3.1.0
+ */
+class Rest {
```
```suggestion
class Rest_Post_Metadata {
```

Given the high likelihood we'll empower additional REST API functionality with this plugin, I suggest we be more specific with class naming.

pauarge replied:
I don't necessarily agree with this change. We will for sure have more REST functionality, which will be spawned from a base `Rest` class. That `Rest` class will then instantiate `Rest_Post_Metadata` or whatever other features we want to implement.

At this point, we don't have other features, so I feel separating those classes is overengineering. When we do implement a second feature, I agree that we should move the specifics of the page/post metadata to its own separate class.

jblz commented on src/class-rest.php:
```diff
@@ -0,0 +1,103 @@
+ * @since 3.1.0
+ */
+class Rest {
+	private const PARSELY_REST_VERSION    = '1.0.0';
+	private const PARSELY_REST_FIELD_NAME = 'parsely';
```
```suggestion
	private const REST_VERSION    = '1.0.0';
	private const REST_FIELD_NAME = 'parsely';
```

If we're more specific about what this class is doing, I don't think we need these prefixes.

jblz commented on src/class-rest.php:
```diff
@@ -0,0 +1,103 @@
+	 *
+	 * @return void
+	 */
+	public function register_parsely_meta(): void {
```
```suggestion
	public function register_meta(): void {
```

jblz commented on src/class-rest.php:
```diff
@@ -0,0 +1,103 @@
+	}
+
+	/**
+	 * Registers the `parsely-meta` field in the REST API.
```
```suggestion
	 * Registers the meta field on the appropriate resource types in the REST API.
```

This would need to be corrected to the proper value anyhow, but since we're sourcing it from a class constant, I think it's appropriate to leave the exact field name out of the doc block.

jblz commented on src/class-rest.php:
```diff
@@ -0,0 +1,103 @@
+	 * @return array The `parsely` object to be rendered in the REST API. Contains a version number describing the
+	 * response and the `meta` object containing the actual metadata.
+	 */
+	public function parsely_rest_get_callback( array $object ): array {
```
I'd just call this `<CLASSNAME>::get_callback`, so it can be more easily ported to additional integrations.  The context is in the namespace and class name.

```suggestion
	public function get_callback( array $object ): array {
```

jblz reviewed (changes requested):
This is working really well.  I left some comments about symbol naming that I think we should address.

pauarge committed:
Update tests/Integration/RestTest.php

Co-authored-by: Jeff Bowen <jblz@users.noreply.github.com>

pauarge committed:
Update src/class-rest.php

Co-authored-by: Jeff Bowen <jblz@users.noreply.github.com>

pauarge committed:
Rename register_parsely_meta to register_meta

pauarge committed:
Renaming parsely_rest_get_callback to get_callback

pauarge committed:
Renaming constants

jblz reviewed (approved):
Nice work <3

pauarge committed:
Merge branch 'develop' into add/decoupled
